[WIP] Telcodocs 85: Validate and Document Intel SRO and SRIOV FEC Operator

kquinn1204 opened the pull request:
Addressing ticket https://issues.redhat.com/browse/CNF-1498 opened incorrectly against the master branch in PR https://github.com/openshift/openshift-docs/pull/31761

Preview build: https://deploy-preview-32785--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-optimize-data-performance-n3000.html

Targetted 4.6

kquinn1204 commented:
Published content is here https://deploy-preview-32785--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-optimize-data-performance-n3000.html

bobfuru commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,248 @@
++
+[NOTE]
+====
+Some tests may be skipped, however the key thing to look out for is the vector tests passing.
```
Could this be tightened a bit? "While some tests can be skipped, be sure that the vector tests pass".

bobfuru reviewed (commented):
Overall, this is looking really good. The info arch is there and you tackle a ton of content very clearly and succinctly. Also, the formatting generally looks good. Many of my comments are based on lack of subject knowledge. (I'm not very familiar with Intel hardware docs, especially in OCP.) I commented that the reserved symbol might only be necessary on first instance, but I saw at the end of my review from your additional resources links (such as [this one](https://catalog.redhat.com/software/operators/detail/6001a748e4e3f23b0b6ad765)) that we use the symbol the same way there. So you might be able to just ignore my suggestion.

Otherwise, many of my comments are based on ISG (it is a LOT to remember!), along with some other OCP docs conventions. The rest are suggestions for readability and clarity.

I hope most of this is helpful. Happy to review again or answer any questions once you've gone through my suggestions. Well done!

bobfuru commented:
Original PR with SME comments here: https://github.com/openshift/openshift-docs/pull/31761

bobfuru commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,156 @@
+
+.. On the *Install Operator* page, select *All namespaces on the cluster*. Then, click *Install*.
+
+. Optional: Verify that the n3000-operator is installed successfully:
```
Or you could just say "Verify that the N3000 Operator is installed successfully"

kquinn1204 commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,173 @@
+
+.Prerequisites
+
+* A cluster installed on bare-metal hardware.
```
I am thinking of leaving this as is as we seem to use bare-metal throughout PAO section in docs https://docs.openshift.com/container-platform/4.7/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#about_hyperthreading_for_low_latency_and_real_time_applications_cnf-master. I thought my lead from these sections where this content logically sits

kquinn1204 commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,156 @@
+
+. Install the N3000 Operator in the namespace you created in the previous step by creating the following objects:
+
+.. Create the following `OperatorGroup` CR and save the YAML in the `n3000-operatorgroup.yaml` file:
```
Yes not particularly sure either was just following an example in existing docs.

mikemckiernan replied:
Please repeat the changes for parallelism that I mentioned in the review for the preceding file,  modules/cnf-installing-sriov-fec-operator.adoc.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,173 @@
++
+[source,terminal]
+----
+$ oc create -f sriov-namespace.yaml
```
Unsolicited feedback...any reason not to use `oc apply` since IIUC based on a recent comment from Vikram in forum-docs, that we have been recommending the use of `oc apply` since forever.

kquinn1204 replied:
I am following along on this guidance here https://docs.openshift.com/container-platform/4.7/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#installing-the-performance-addon-operator_cnf-master and modelled this section on that section. I did read your discussion on oc apply. I meant to bring it up at the team meeting but from our docs, there seems to be a lot of `oc apply` instructions. So I followed the same when it comes to applying the CR. Here I am creating the `Namespace` and `OperatorGroup` CR  later I come to apply the CR that defines the test and I use `oc apply`.

jboxman replied:
@mikemckiernan, this is basically the same pattern for Operator installation used everywhere. I've updated the SR-IOV Network Operator variant to use here documents and so forth, but it never caught on I guess.

kquinn1204 commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,248 @@
+sh-4.4# export CPU=$(cat /sys/fs/cgroup/cpuset/cpuset.cpus)
+sh-4.4# echo ${CPU}
+----
+This prints out the CPUs assigned to the fec.pod.
```
I don't think it needs that clarification so I decided to leave as This prints out the CPUs that are assigned to the `fec.pod`

jboxman commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,172 @@
++
+[source,terminal]
+----
+$ oc get packagemanifest sriov-operator -n openshift-marketplace -o jsonpath='{.status.defaultChannel}'
```
@kquinn1204, is the `defaultChannel` always correct? For SR-IOV Network Operator, the `channel` is based on the OCP version, so this is handled a bit differently, because `defaultChannel` is not necessarily correct.

SchSeba replied:
here the channel is always stable for the operator

kquinn1204 replied:
Thanks, @SchSeba I was not sure on this one. Hopefully, a QE will bottom out any other technical issues

jboxman commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,349 @@
+
+.Procedure
+
+. Verify that the N3000 Operator is installed and that the pods are running:
```
@kquinn1204 are certain permissions required to run this?

SchSeba replied:
I think is a cluster-admin the only one can apply this yamls

kquinn1204 replied:
I added "Log in as a user with cluster-admin privileges" to Prerequisites

jboxman commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,349 @@
+N3000-discovery-blmjl                       1/1         Running         1           15d
+N3000-discovery-lblh7                       1/1         Running         1           15d
+----
+* `fpga-driver-daemonset` provides and loads the required Open Programmable Accelerator Engine (OPAE) drivers
```
@kquinn1204, I wonder if including an introduction sentence makes sense here?

```
+
--
In the previous output, the following...

* a
* b
--
```

jboxman commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,349 @@
+
+. Update the user bitstream of the Intel® FPGA PAC N3000 card user:
+
+.. Create a custom resource (CR) containing information about what node and card should be programmed together with the bitstream you want to flash the card with. Save the `YAML` in the `n3000-cluster.yaml` file.
```
@kquinn1204, we've been omitting code font for `YAML`, so just "YAML".

SchSeba commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,349 @@
+status:
+      conditions:
+          - lastTransitionTime: "2020-12-15T18:18:53Z"
+          message: Flashed successfully <1>
```
indetation issue here everything needs to be on the same line as `lastTransitionTime`

SchSeba commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,344 @@
+      - deviceID: 0d5c <1>
+        driver: pci-pf-stub
+        maxVirtualFunctions: 16
+            pciAddress: 0000:af:00.0
```
indentation issue needs to be in the same line as  `maxVirtualFunctions` all the lines below also

SchSeba commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+status:
+    conditions:
+        - lastTransitionTime: "2021-03-19T17:19:37Z"
+        message: Configured successfully
```
indentation issue all the lines need to be in the `lastTransitionTime` line

SchSeba commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+                n3000:
+                  # Network Type: either "FPGA_5GNR" or "FPGA_LTE"
+                  networkType: "FPGA_5GNR"
+                  # Programming mode: 0 = VF Programming, 1 = PF Programming
```
lets remove this comment here we only support VF mode so pfMode should always be false

SchSeba commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+      type: Configured
+    inventory:
+      sriovAccelerators:
+    - deviceID: 0d8f <1>
```
indentation issue the `-` should start as `sriovAccelerators` all the lines below should move accordingly

SchSeba commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,251 @@
+        add:
+          - IPC_LOCK
+          - SYS_RESOURCE
+    resources:
```
indentation issue should move to the `capabilities` level

kquinn1204 replied:
Hi @SchSeba I was checking against https://github.com/open-ness/openshift-operator/blob/main/spec/openshift-sriov-fec-operator.md and it looks like resources is on the same level as securityContext. You want me to put at the same level as capabilities  and indent others accordingly - correct?

capabilities:
        add:
          - IPC_LOCK
          - SYS_RESOURCE
      resources:
       requests:
         hugepages-1Gi: 4Gi <4>
         memory: 1Gi
         cpu: "4" <5>
         intel.com/intel_fec_5g: '1' <6>
         #intel.com/intel_fec_acc100: '1'
         #intel.com/intel_fec_lte: '1'
       limits:
         memory: 4Gi
         cpu: "4"
         hugepages-1Gi: 4Gi
         intel.com/intel_fec_5g: '1'
         #intel.com/intel_fec_acc100: '1'
         #intel.com/intel_fec_lte: '1'

SchSeba commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,251 @@
+====
+To test the ACC100 configuration, uncomment `intel.com/intel_fec_acc100` by removing the # symbol.
+To test the N3000 4G/LTE configuration, uncomment `intel.com/intel_fec_lte` by removing the # symbol.
+Only one test configuration should be active at any time.
```
"test configuration" -> resources

SchSeba reviewed (commented):
great job, almost there from my side just some indentation issues

SchSeba commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,251 @@
+spec:
+  containers:
+  - securityContext:
+      privileged: false
```
on the open-ness doc the indentation is right 

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-bbdev-sample-app
spec:
  containers:
  - securityContext:
      privileged: false
      capabilities:
        add:
          - IPC_LOCK
          - SYS_NICE
    name: bbdev-sample-app
    image: bbdev-sample-app:1.0
    command: [ "sudo", "/bin/bash", "-c", "--" ]
    args: [ "while true; do sleep 300000; done;" ]
    volumeMounts:
    - mountPath: /hugepages
      name: hugepage
    resources:
      requests:
        intel.com/intel_fec_acc100: '1'
        hugepages-1Gi: 2Gi
        memory: 2Gi
      limits:
        intel.com/intel_fec_acc100: '1'
        hugepages-1Gi: 2Gi
        memory: 2Gi
  volumes:
  - name: hugepage
    emptyDir:
      medium: HugePages
```

SchSeba commented:
this looks great just my previews command on the pod definition and you have a lgtm after that thanks!

jboxman commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+[id="cnf-understanding-intel-hardware-accelertor-card-for-openshift-container-platform_{context}"]
+= Understanding Intel hardware accelerator cards for OpenShift Container Platform
+
+Hardware accelerator cards from Intel®  play a key role in accelerating 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads, which in turn increases the overall compute capacity of a commercial, off-the-shelf platform.
```
```suggestion
Hardware accelerator cards from Intel® play a key role in accelerating 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads, which in turn increases the overall compute capacity of a commercial, off-the-shelf platform.
```

jboxman commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+The Intel® FPGA PAC N3000 is a full-duplex, 100 Gbps in-system, re-programmable acceleration card for multi-workload networking application acceleration. It has an optimal memory mixture designed for network functions, with an integrated network interface card (NIC) in a small form factor that enables high throughput, low latency, and low power per bit for a custom networking pipeline.
+
+When the Intel® FPGA PAC N3000 is programmed with a 4G/LTE or 5G bitstream, it exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices which can be used to accelerate the FEC in the vRAN workload.
+To take advantage of this functionality for a cloud-native deployment, the physical function (PF) of the device must be bound to the must be bound to `pf-pci-stub` driver to create several VFs. Once the VFs are created, they must also be bound to a DPDK userspace driver (vfio) to allocate them to specific pods running the vRAN workload.
```
```suggestion
To take advantage of this functionality for a cloud-native deployment, the physical function (PF) of the device must be bound to the `pf-pci-stub` driver to create several VFs. Once the VFs are created, they must also be bound to a DPDK userspace driver (vfio) to allocate them to specific pods running the vRAN workload.
```

jboxman commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+* Intel FPGA PAC N3000 card
+* Performance Addon Operator with RT kernel configuration
+* Node or nodes installed with the N3000 Operator
+* Log in as a user with cluster-admin privileges
```
```suggestion
* Log in as a user with `cluster-admin` privileges
```

jboxman commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
++
+[NOTE]
+====
+All the commands run in the `vran-acceleration-operators` namespace.
```
Of their own accord, or does this require `-n vran-acceleration-operators` as part of the `oc` commands?

kquinn1204 replied:
good question @jboxman I was taking my lead from u/s docs https://github.com/open-ness/openshift-operator/blob/main/spec/openshift-pacn3000-operator.md @SchSeba can you confirm the correct syntax do I need to specify -n vran-acceleration-operators where I state all commands run in the vran-acceleration-operators namespace

jboxman commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,344 @@
+Without FEC, a garbled message would have to be resent, adding to the network load and impacting both throughput and latency.
+====
+
+.Prerequisites
```
As `cluster-admin`?

jboxman commented on scalability_and_performance/cnf-optimize-data-performance-n3000.adoc:
```diff
@@ -0,0 +1,25 @@
+
+include::modules/cnf-intel-n3000-optimize-data-plane.adoc[leveloffset=+1]
+
+.Additional resources
```
At the assembly level, we've been doing this as a header

```
[id="{context}-additional-resources"]
== Additional resources
```

jboxman reviewed (commented):
I left a few comments for your review; I work adjacent to this, so I've been curious what this hardware does.

SchSeba reviewed (commented):
DEV ACK

evgenLevin commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,251 @@
++
+[source,terminal]
+----
+$  oc get pods
```
`pod-bbdev-sample-app` was created in namespace `test-bbdev`, but here we check `vran-acceleration-operators` namespace

evgenLevin commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,251 @@
+[source,terminal]
+----
+NAME                                            READY           STATUS          RESTARTS        AGE
+fec-pod                                         1/1             Running         0               10s
```
the pod name should be `pod-bbdev-sample-app` not `fec-pod` according to` pod-test.yaml`

evgenLevin commented:
QA ACK

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+
+One of the most compute-intensive 4G/LTE and 5G workloads is RAN layer 1 (L1) forward error correction (FEC). FEC resolves data transmission errors over unreliable or noisy communication channels. FEC technology detects and corrects a limited number of errors in 4G/LTE or 5G data without the need for retransmission.
+
+The FEC devices are provided by the Intel® FPGA PAC N3000 FPGA and the Intel® vRAN Dedicated Accelerator ACC100 for the vRAN use case.
```
Please remove the trademark symbol.  The text "Intel" has been used many, many times in the product documentation without incident.

kquinn1204 replied:
I think trademark relevant as in catalog trademark is there OpenNESS Operator for Intel® FPGA PAC N3000 so when installing etc. I am going to retain it. I other cases I will remove it. I think that is OK

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+
+[NOTE]
+====
+The Intel® FPGA PAC N3000 FPGA requires flashing with an 4G/LTE or 5G bitstream.
```
Is this temporary or permanent? Does Intel provide documentation that describes how to do this?

Just making a check that you have what you want--you have FPGA in there twice.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+The Intel® FPGA PAC N3000 FPGA requires flashing with an 4G/LTE or 5G bitstream.
+====
+
+The OpenNESS SR-IOV Operator for Wireless FEC Accelerators provides functionality to create desired virtual functions (VFs) for the FEC device, binds them to appropriate drivers, and configures the VFs queues for desired functionality in 4G/LTE or 5G deployment.
```
Remove both instances of "desired."

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+
+. Install the OpenNESS SR-IOV Operator for Wireless FEC Accelerators in the namespace you created in the previous step by creating the following objects:
+
+.. Create the following `OperatorGroup` CR and save the YAML in the `sriov-operatorgroup.yaml` file:
```
For parallelism, why isn't this instruction like the previous one?

```
Define the `vran-operators` operator group by creating a file named `sriov-operatorgroup.yaml` as shown in the following example:
```

kquinn1204 replied:
I am following the same structure I observed in the existing documentation regarding how to create the SR-IOV operator.

mikemckiernan replied:
Thanks for letting me know the origin.  I'll put that on my list to fix.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+      - vran-acceleration-operators
+----
+
+.. Create the `OperatorGroup` CR by running the following command:
```
s/`OperatorGroup` CR/operator group/

That matches the previous set of commands.

kquinn1204 replied:
I am following the same structure as shown here https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html which I have stepped through on my local machine and made sense

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+stable
+----
+
+.. Create the following `Subscription` CR and save the YAML in the `sriov-sub.yaml` file:
```
Revise to be parallel with the preceding instrucitons or change the first instruction to be parallel with this one and the preceding instruction.  My $0.02 is that omitting the backticks and abbreviations aids readability.

```
Define the subscription for the operator by creating a file named `sriov-sub.yaml` as shown in the following example:
```

kquinn1204 replied:
I am following the same structure as shown here https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html which I have stepped through on my local machine and made sense

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+$ oc create -f sriov-sub.yaml
+----
+
+.. Verify that the Operator is installed:
```
Add a `.Verification` block title and use a single bullet instead of a number for the step.

https://redhat-documentation.github.io/modular-docs/#procedure-module-guidelines

mikemckiernan replied:
Please change the ".." to "*" so that customers see a single bullet instead of "a.".

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+
+[NOTE]
+====
+You must create the `Namespace` and `OperatorGroup` CR as mentioned in the previous section.
```
Bummer.

jboxman replied:
@mikemckiernan, so this came up for SR-IOV Network Operator as well, and I _think_ there is a way to do this in the web console at this point, but I never had time to go back and make the changes; this must have been 9 months ago now...

kquinn1204 replied:
Yes you can do this all in Web Console and I once again was trying to be consistent with another existing documentation section - https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#installing-the-performance-addon-operator_cnf-master which for low latency use cases "Telco edge" would also need to be followed

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,157 @@
+
+The OpenNESS Operator for Intel® FPGA PAC N3000 orchestrates and manages the resources or devices exposed by the Intel® FPGA PAC N3000 card within the {product-title} cluster.
+
+For vRAN use cases the OpenNESS Operator for Intel® FPGA PAC N3000 is used alongside the OpenNESS Operator for Intel® Wireless FEC Accelerator.
```
Can you add a comma after "use cases"?

s/alongside/with/

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,157 @@
+[source,terminal]
+----
+$ oc project vran-acceleration-operators
+
```
Please remove blank line.

kquinn1204 replied:
I get your point when I started this write up I was following the guidance in the u/s docs which indicated this is the way it is done. I guess I was ensuring steps broadly followed that described upstream

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+
+.. Create the following `Subscription` CR and save the YAML in the `sriov-sub.yaml` file:
++
+.Example Subscription
```
You can remove the `.Example subscription` block title.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+----
++
+.Example output
++
```
IIRC, no + between the block title and the block that follows.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="installing-the-sriov-fec-operator_{context}"]
```
I doubt that the civilization crumbles if the ID doesn't *exactly* match the module file name, but it's written down.  See #2:

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,157 @@
+
+.. Create the following `Subscription` CR and save the YAML in the `n3000-sub.yaml` file:
++
+.Example Subscription
```
Omit the block heading--along with the identical suggestions from my review of the preceding file.

mikemckiernan commented on modules/cnf-installing-sriov-fec-operator.adoc:
```diff
@@ -0,0 +1,170 @@
+During installation an Operator might display a *Failed* status. If the installation later succeeds with an *InstallSucceeded* message, you can ignore the *Failed* message.
+====
++
+If the Operator does not appear as installed, to troubleshoot further:
```
Maybe "If the console does not indicate that the Operator is installed, perform the following troubleshooting steps:"

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+[id="cnf-understanding-intel-hardware-accelertor-card-for-openshift-container-platform_{context}"]
+= Understanding Intel hardware accelerator cards for OpenShift Container Platform
+
+Hardware accelerator cards from Intel® accelerates 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads, increasing in turn the overall compute capacity of a commercial, off-the-shelf platform.
```
s/accelerates/accelerate/, right?

I also think you have two sentences.  I'm leery of suggesting an alternative because I am unsure how the hardware processes packets to increase the compute capacity.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+
+Hardware accelerators cards from Intel®  play a key role in accelerating 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads, which in turn increases the overall compute capacity of a commercial, off-the-shelf platform.
+
+The hardware accelerator cards detailed in this section are:
+
+* Intel® FPGA PAC N3000
+
+* Intel® vRAN Dedicated Accelerator ACC100
```
Please remove lines 11 to 15 as Bob suggested, for the same reason.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
++
+[NOTE]
+====
+The Intel® FPGA PAC N3000 card needs flashed with the 4G/LTE or 5G bitstreams to support vRAN workloads.
```
This can be the first paragraph under the `.Intel FPGA PAC N3000` block title.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+
+The Intel® FPGA PAC N3000 is a reference FPGA and uses 4G/LTE or 5G Forward Error Correction (FEC) as an example workload accelerating the 5G or 4G/LTE RAN layer 1 (L1) base station network function.
+
+The Intel® FPGA PAC N3000 is a full-duplex, 100 Gbps in-system, re-programmable acceleration card for multi-workload networking application acceleration. It has an optimal memory mixture designed for network functions, with an integrated network interface card (NIC) in a small form factor that enables high throughput, low latency, and low power per bit for a custom networking pipeline.
```
How much is Intel paying us?  I think you can keep it to 100 Gpbs and that it is reprogrammable (the FP in FPGA), and then expect that the customer can find in-depth information about the hardware from the Intel website.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+The Intel® FPGA PAC N3000 is a full-duplex, 100 Gbps in-system, re-programmable acceleration card for multi-workload networking application acceleration. It has an optimal memory mixture designed for network functions, with an integrated network interface card (NIC) in a small form factor that enables high throughput, low latency, and low power per bit for a custom networking pipeline.
+
+When the Intel® FPGA PAC N3000 is programmed with a 4G/LTE or 5G bitstream, it exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices used to accelerate the FEC in the vRAN workload.
+To take advantage of this functionality for a cloud-native deployment, the physical function (PF) of the device must be bound to the `pf-pci-stub` driver to create several VFs. Once the VFs are created, they must also be bound to a DPDK userspace driver (vfio) to allocate them to specific pods running the vRAN workload.
```
s/Once/After/

s/they must also be/they must be/

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+
+Intel® FPGA PAC N3000 support on {product-title} depends on two Operators:
+
+. OpenNESS Operator for Intel® FPGA PAC N3000 (Programming)
```
Use `*` to create bullets instead of an ordered list.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+.vRAN Dedicated Accelerator ACC100
+
+The vRAN Dedicated Accelerator ACC100, based on Intel’s eASIC technology is designed to offload and accelerate the computing-intensive process of forward error correction for 4G/LTE and 5G technology, freeing up processing power.
+Intel eASIC™ devices are structured ASICs, an intermediate technology between FPGAs and standard Application-Specific Integrated Circuits (ASICs).
```
Please remove the trademark symbol.  Lower case "application-specific integrated circuit."  They are special, but not that special.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,45 @@
+The vRAN Dedicated Accelerator ACC100, based on Intel’s eASIC technology is designed to offload and accelerate the computing-intensive process of forward error correction for 4G/LTE and 5G technology, freeing up processing power.
+Intel eASIC™ devices are structured ASICs, an intermediate technology between FPGAs and standard Application-Specific Integrated Circuits (ASICs).
+
+These devices provide lower unitcost and lower power compared to FPGAs and faster time to market and lower non-recurring engineering cost compared to standard- ASICs.
```
s/unitcost/unit cost/

I think you can remove the hyphen after "standard."  FWIW, I doubt that the stated benefits are of interest to the administrative audience for this product documentation.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="programming-the-openness-operator-for-intel-fpga-pac-n3000-operator_{context}"]
```
Like previous comment about the ID, it's cheap to make it match the file name exactly.

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+[id="programming-the-openness-operator-for-intel-fpga-pac-n3000-operator_{context}"]
+= Programming the OpenNESS Operator for Intel® FPGA PAC N3000
+
+This section describes how to program the Intel® FPGA PAC N3000 with a vRAN 5G bitstream. This bitstream exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices used to accelerate the FEC in the vRAN workload.
```
Instead of the self-referential "This section...", I think you can use "As an administrator, you can program the Intel FPGA PAC..."

s/devices used/devices that are used/

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+
+* Intel FPGA PAC N3000 card
+* Performance Addon Operator with RT kernel configuration
+* Node or nodes installed with the N3000 Operator
```
Is the N3000 Operator the same as the OpenNESS <lots-of-other-words> Operator?  If yes, then just repeat the same lengthy term that customers can find in a heading somewhere else for the purpose of findability.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+. Verify that the N3000 Operator is installed and that the pods are running:
++
+[source,terminal]
+----
```
Some speculation:

* This step appears to be the missing verification step from the installation procedure.
* The mysterious `oc project ...` command that the installation procedure ended with is an explanation for why these commands do not require the `-n vran-acceleration-operators` namespace argument that Jason asked about.

I _am_ speculating, but those pieces fit together tightly.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,349 @@
+----
+* `fpgainfo-exporter` provides N3000 telemetry data for Prometheus
+* `fpga-driver-daemonset` provides and loads the required Open Programmable Accelerator Engine (OPAE) drivers
+* `N3000-controller-manager` applies N3000Node CRs to the cluster and manages all the operand containers
```
s/all// too, please.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+* `fpga-driver-daemonset` provides and loads the required Open Programmable Accelerator Engine (OPAE) drivers
+* `fpgainfo-exporter` provides N3000 telemetry data for Prometheus
+* `N3000-controller-manager` applies N3000Node CRs to the cluster and manages all the operand containers
+* `N3000-daemonset` is the main worker application
```
"main worker application" is imprecise.  Is this the software that exposes the VFs?  If yes, then please state that instead.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+node1                      NotRequested
+----
+
+. Retrieve information about the card on each node:
```
s/Retrieve/Get/ for consistency.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+    bitstreamVersion: 0.2.3
+    deviceId: "0x0b30"
+----
+<1> Specify the PCI address of the card to be programmed.
+<2> This is the currently flashed bitstream.
```
<1> The `PCIAddr` field indicates the PIC address of the card.
<2> The `bitstreamId` field indicates the bitstream that is currently stored in flash.

Bit of a guess on those, but please revise so they are parallel. Our customer is not specifying anything at this point--just viewing the result of the preceding command.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+# oc get n3000node -o json
+----
+
+. Update the user bitstream of the Intel® FPGA PAC N3000 card user:
```
Just a check--is "card user" at the end of this instruction correct?

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+
+. Update the user bitstream of the Intel® FPGA PAC N3000 card user:
+
+.. Create a custom resource (CR) containing information about what node and card should be programmed together with the bitstream you want to flash the card with. Save the YAML in the `n3000-cluster.yaml` file.
```
Using the pattern from the first instruction, I think you can snip this down:

```
Define the N3000 cluster resource to program by creating a file named `n3000-cluster.yaml` as shown in the following example:
```

All the details are spelled out in callouts so I think you can leave the instructional text a little more spare.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+            PCIAddr: "0000:1b:00.0" <4>
+            checksum: "0b0a87b974d35ea16023ceb57f7d5d9c" <5>
+----
+<1> This name is mandatory and is the only one that can be used.
+<2> Specify the node to be programmed.
+<3> Specify the URL for the user bitstream. This bitstream file needs to be accessible on an HTTP or HTTPS server.
+<4> PCI address of the card to be programmed.
+<5> This is the MD5 checksum of the bitstream requested under the `userImageURL`. Use this for checksum verification.
```
This YAML looks very helpful.  I think the callouts can be more parallel and consistent.

<1> Specify the name. The name must be `n3000`.
<2> Specify the node to program.
<3> what you have, but s/needs to be/must be/.
<4> Specify the PCI address of the card to program.
<5> Specify the MD5 checksum of the bitstream that is specified in the `userImageURL` field.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+[NOTE]
+====
+The N3000 daemon updates the FPGA user bitstream using the Open Programmable Acceleration Engine (OPAE) tools and resets the PCI device.
+The update of the FPGA user bitstream may take up to 40 minutes per card. For programming cards on multiple nodes, the programming happens one node at a time.
```
s/may take/can require/

s/the programming happens/the programming is performed/

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
++
+[source,terminal]
+----
+# oc get n3000node -o json
```
I can't connect this step that gets JSON to the next step begins by creating a YAML file.

If you feel like you need to keep it, the prompt is `$` instead of `#`.

kquinn1204 replied:
I think it just provides a better-formatted output. AFAIR it was used on a demo of this process.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+The N3000 daemon updates the FPGA user bitstream using the Open Programmable Acceleration Engine (OPAE) tools and resets the PCI device.
+The update of the FPGA user bitstream may take up to 40 minutes per card. For programming cards on multiple nodes, the programming happens one node at a time.
+====
+.. Apply the CR.
```
End the instructional text with a colon and maybe something like "Apply the update to begin programming the card with the bitstream:"

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
++
+[source,terminal]
+----
+$ oc get pod -o wide | grep n3000-daemonset | grep node1
```
This would be a little more polished to get the pods by label with `oc get pod -l app=n3000-daemonset`.

Can we get an SME to confirm that label is accurate?  I'm guessing from the foll URL:

https://github.com/open-ness/openshift-operator/blob/ca85a7659d695c5e006b4372d9ef993875e235ef/N3000/assets/400-daemon.yaml#L113

kquinn1204 replied:
I think you are correct. I can reach out to QE. What is currently in there was checked.

kquinn1204 replied:
@SchSeba Can you confirm if that command would work?

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+----
++
+.Example output
+
```
Remove blank line, according to team guidelines.

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+...
+----
++
+The log file indicates the following flow of events:
```
Very nice.  This is awesome.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+        deviceId: "0x0b30 <4>
+----
++
+<1> The message field indicates the device is successfully flashed.
+<2> Use this field to check that the correct PCI address of the card is programmed.
+<3> Use this field to verify that the `bitstreamId` is changed. Check this against the output obtained from step 5.
+<4> The current deviceID of the bitstream inside the card exposed to the system.
```
<1> The `message` field...
<2> The `PCIAddr` field indicates the PCI address of the card.
<3> The `bitstreamId` field indicates the updated bitstream ID.
<4> The `deviceId` field indicates that device ID of the bitstream...

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+      - PCIAddr: 0000:1b:00.0 <2>
+        bitstreamId: "0x2315842A010601" <3>
+        bitstreamVersion: 0.2.3
+        deviceId: "0x0b30 <4>
```
nit: The second quotation mark is missing.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+** The bitstream is applied.
+* After flashing is complete the PCI device or devices on the node or nodes are reloaded. The Intel wireless Operator is now able to find the new flashed device or devices.
+
+. Verify the status after the FPGA user bitstream update is complete:
```
Please add a `.Verification` block title before this step.  These are all verification steps.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+
+. Check the FEC PCI devices on the node:
+
+.. Ensure the node configuration is applied correctly:
```
s/Ensure/Verify/ for consistency.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+----
+$ oc debug node/node1
+----
+.Expected output
```
Add a blank line above the block title.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+
+sh-4.4#
+----
+.. Verify that you can use the node file system:
```
Add a blank line before this step.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+----
+1b:00.0 Processing accelerators: Intel Corporation Device 0b30
+1d:00.0 Processing accelerators: Intel Corporation Device 0d8f (rev 01)
+
```
Remove blank line from the expected output and remove the blank on line 344.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,351 @@
+1d:00.0 Processing accelerators: Intel Corporation Device 0d8f (rev 01)
+
+----
+Devices belonging to the FPGA are reported in the output, where Device ID `0b30` is the RSU interface used to program the card, and the `0d8f` is a physical function of the newly programmed 5G device.
```
Add a blank line above, use + if you want to keep the text with the step.  Omit the + if this paragraph is a summary of the verification.

My sugg is to replace the first comma with a period and make two sentences.  You can remove "where".

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="configuring-the-sr-iov-fec-operator-intel-vran-dedicated-accelerator-acc100_{context}"]
```
Please match the ID to the file name, according to team guidelines.

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+[id="configuring-the-sr-iov-fec-operator-intel-vran-dedicated-accelerator-acc100_{context}"]
+= Configuring the SR-IOV-FEC Operator for the Intel® vRAN Dedicated Accelerator ACC100
+
+This section describes how to program the Intel® vRAN Dedicated Accelerator ACC100.
```
Please replace this sentence with information that states the benefit of programming the card.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
++
+[NOTE]
+====
+All the commands run in the `vran-acceleration-operators` namespace.
```
Remove this note and add a new step 1 with `oc project vran-acceleration-operators`.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+
+.Procedure
+
+. Verify that the SR-IOV-FEC Operator is installed and that pods are running:
```
Snip it because the next step checks the pods.

"Verify that the SR-IOV-FEC Operator is installed:"

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+----
+* `sriov-device-plugin` expose the FEC virtual functions as resources under the node
+* `sriov-fec-controller-manager` applies CR to the node and maintains the operands containers
+* `sriov-fec-daemonset` is the main worker application
```
Please indicate what the daemon set does rather than "main worker application."

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+* `sriov-fec-controller-manager` applies CR to the node and maintains the operands containers
+* `sriov-fec-daemonset` is the main worker application
++
+[NOTE]
```
Remove the note. If the information is that the example output from the preceding command shows the pods in the `sriov-fec` namespace, then set step 1 to `oc project sriov-fec`.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+----
+
+. Find the physical function (PF) of the SR-IOV FEC accelerator device to be configured:
+
```
Please remove the blank line.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+<1> Specify a name for the CR object. The only name that can be specified is `config`.
+<2> Specify the node name.
+<3> Specify the PCI address of the card on which the SR-IOV-FEC Operator will be installed.
+<3> Specify the number of virtual functions to create. For the Intel® vRAN Dedicated Accelerator ACC100, create all 16 VFs.
```
<4> ? and where in the YAML does that go?

+1 on the parallelism.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+
+.  Configure the FEC device with the desired setting.
+
+.. Create the following custom resource (CR) and save the YAML in the `sriovfec_acc100cr.yaml` file.
```
End the instructional text with a colon.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+Each configured VF has access to all the queues. Each of the queue groups have a distinct priority level. The request for a given queue group is made from the application level that is, the vRAN application leveraging the FEC device.
+====
+
+.. Apply the CR.
```
End the instructional text with a colon.

Ideally, you would also indicate what is configured by applying the YAML.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
++
+After creation of the CR, the SR-IOV FEC daemon starts configuring the FEC device.
+
+.. Check the status:
```
Add a `.Verification` block title before these verification steps.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+        - deviceID: 0d5d
+          driver: vfio-pci
+          pciAddress: 0000:b0:00.4
+----
```
Please repeat the callouts like you did for the previous example.  Customers should check the `message` field for "Configured successfully" and so on, right?

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
++
+[source,terminal]
+----
+$ oc get po -o wide | grep sriov-fec-daemonset | grep node01
```
Ideally, an SME will confirm that you can replace with `oc get pods -l app=sriov-fec-daemonset`.

I'm picking that from the foll:

https://github.com/open-ness/openshift-operator/blob/ca85a7659d695c5e006b4372d9ef993875e235ef/sriov-fec/assets/300-daemon.yaml#L113

kquinn1204 replied:
@SchSeba Can you verify if that command would work?

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+----
++
+.Example output
+
```
Please remove blank line and add one before the next step.

mikemckiernan replied:
Please again.

mikemckiernan replied:
Please again.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+----
++
+.Example output
+
```
Please remove the blank line.

mikemckiernan replied:
Please again.

mikemckiernan replied:
Please again.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+{"level":"Level(-4)","ts":1616794346.8040674,"logger":"daemon.updateInventory","msg":"obtained inventory","inv":{"sriovAccelerators":[{"vendorID":"8086","deviceID":"0b32","pciAddress":"0000:20:00.0","driver":"","maxVirtualFunctions":1,"virtualFunctions":[]},{"vendorID":"8086","deviceID":"0d5c","pciAddress":"0000:af:00.0","driver":"pci-pf-stub","maxVirtualFunctions":16,"virtualFunctions":[{"pciAdd ...
+{"level":"Level(-4)","ts":1616794346.9058325,"logger":"daemon","msg":"Update ignored, generation unchanged"}
+{"level":"Level(-2)","ts":1616794346.9065044,"logger":"daemon.Reconcile","msg":"Reconciled","namespace":"vran-acceleration-operators","name":"pg-itengdvs02r.altera.com"}
+----
```
Bummer that customers don't get the highlights of what to look for like you showed with the preceding instance.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,345 @@
+----
+
+. Check the FEC configuration of the card:
+
```
Remove blank line.

mikemckiernan replied:
Please again.

mikemckiernan replied:
Please again.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="configuring-the-sr-iov-fec-operator-intel-fpga-pac-n3000_{context}"]
```
Please match the ID to the file name.

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+* Intel® FPGA PAC N3000 card
+* Node or nodes installed with the OpenNESS Operator for Intel® FPGA PAC N3000 (Programming)
+* Node or nodes installed with the OpenNESS Operator for Wireless FEC Accelerators
+* RT kernel configured with Performance Addon Operator
```
I guess Jason ran out of steam, so I'll ask, `cluster-admin`?

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+* RT kernel configured with Performance Addon Operator
++
+[NOTE]
+====
```
As with other instances, please delete this note and set the first step to `oc project vran-acceleration-operators`, assuming that is the correct namespace.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+
+.Procedure
+
+. Verify that the SR-IOV-FEC Operator is installed and that pods are running:
```
This instructional text is identical to the text in the preceding file.  Curiously, the command in that file specified a namespace and this `oc` command instance does not.

If the instructional text is correct, please snip it to "Verify that the SR-IOV-FEC Operator is installed:"

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+* `N3000-discovery` discovers N3000 Accelerator devices installed and labels worker nodes if devices are present
+* `sriov-device-plugin` expose the FEC virtual functions as resources under the node
+* `sriov-fec-controller-manager` applies CR to the node and maintains the operands containers
+* `sriov-fec-daemonset` is the main worker application
```
Ideally, you can repeat the more precise description from other files and replace "main worker application" in these two instances.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
++
+[NOTE]
+====
+This output shows the `N3000` and the `sriov-fec` FEC pods.
```
Remove the note and update the instructional text from "Verify that the pods are running:" to include `N3000` and `sriov-fec`.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+This output shows the `N3000` and the `sriov-fec` FEC pods.
+====
+
+
```
Remove blank line.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+----
+
+. Find the physical function (PF) of the SR-IOV FEC accelerator device to be configured:
+
```
Remove blank line.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+
+.  Configure the FEC device with the desired setting.
+
+.. Create the following custom resource (CR) and save the YAML in the `sriovfec_n3000_cr.yaml` file.
```
Ideally, the instructional text would follow the pattern that you established in the first file.

End the instructional text with a colon.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+<2> Specify the PCI Address of the card on which the SR-IOV-FEC Operator will be installed.
+<3> Specify the number of virtual functions. Create two virtual functions.
+<4> On `vf0` create one queue with 16 buses (downlink and uplink).
+<5> On `vf1` create one queue with 16 buses (downlink and uplink).
```
For 4 & 5, "Specify one queue on `vfx` with 16 buses..."

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+====
+For Intel® PAC N3000 for vRAN Acceleration the user can create up to 8 VF devices. Each FEC PF device provides a total of 64 queues to be configured, 32 queues for uplink and 32 queues for downlink. The queues would be typically distributed evenly across the VFs.
+====
+.. Apply the CR.
```
Add a blank line before step.

End the instructional text with a colon.  Indicate what is configured by applying the YAML.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
++
+After creation of the CR, the SR-IOV FEC daemon starts configuring the FEC device.
+
+.. Check the status:
```
Add a `.Verification` block title before this step.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+----
++
+.Example output
+
```
Remove blank line.

mikemckiernan replied:
Please again.

mikemckiernan replied:
Please again.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+----
+sriov-fec-daemonset-kqqs6                      1/1     Running   0          19h
+----
+.. View the logs:
```
Add a blank line before the step.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
++
+[source,terminal]
+----
+$ oc get pod | grep sriov-fec-daemonset
```
If an SME approves selecting the pods in the `sriov-fec-daemonset` by label, please make the same update here.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+----
++
+.Example output
+
```
Remove blank line.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+2020-12-16T12:46:48.160Z        INFO    daemon.drainhelper.Run()        worker function - end   {"performUncordon": true}
+----
+
+. Check the FEC configuration of the card.
```
End instructional text with a colon.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,398 @@
+        pciAddress: 0000:1d:00.2
+----
+<1> `0d8f` is the `deviceID` physical function of the FEC device
+<2> `0d90` is the `deviceID` virtual function of the FEC device
```
These are good.  Please add a period at the end of each.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="verify-application-pod-access-and-fpga-usagage-on-openess_{context}"]
```
Please set the ID to match the file name.

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+[id="verify-application-pod-access-and-fpga-usagage-on-openess_{context}"]
+= Verifying application pod access and FPGA usage on OpenNESS
+
+OpenNESS is an edge computing software toolkit enabling edge platforms to onboard and manage applications and network functions on any type of network.
```
s/toolkit enabling/toolkit that enables/

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,248 @@
+
+.Prerequisites
+
+* Intel® FPGA PAC N3000 card (Optional)
+* Node or nodes installed with the n3000-operator
+* Node or nodes installed with the SR-IOV-FEC operator
+* Real-Time kernel and huge pages configured with Performance Addon Operator
```
Please use Optional: as Bob indicated and his other edits too.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+    openshift.io/run-level: "1"
+----
+
+.. Create the namespace by running the following command:
```
Excellent.  You didn't indicate to apply the CR and instead indicated what the customer accomplishes.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,248 @@
+<1> Specify the `namespace` you created in step 1.
+<2> This is a test image containing the compiled DPDK.
+<3> Make the container execute internally as the root user.
+<4> Specify hugepage size hugepages-1Gi and the quantity of hugepages that will be allocated to the pod. Hugepages and isolated CPUs needs to be configure using the Performance Addon Operator.
```
There is subject and verb disagreement in the second sentence. Please revise as Bob suggested.

mikemckiernan replied:
Please revise the second sentence.

s/using/by using/.  Check ISG entry for "using" around p427.

mikemckiernan replied:
Please review the preceding comment.  On this third reading, I suspect that you want the first sentence to identify the huge page size as a literal: "Specify huge page size `hugepages-1Gi` and..."

Literal values are enclosed in backticks as mentioned in the [quick reference](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference) section of the team guidelines.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+====
+To test the ACC100 configuration, uncomment `intel.com/intel_fec_acc100` by removing the # symbol.
+To test the N3000 4G/LTE configuration, uncomment `intel.com/intel_fec_lte` by removing the # symbol.
+Only one resource should be active at any time.
```
Revise so that "should" does not appear in the sentence.  "can be active" might be an alternative if it is technically accurate.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+pod-bbdev-sample-app                            1/1             Running         0               80s
+----
+
+. Remote shell log in to the `pod-bbdev-sample-app`:
```
"Remote shell" isn't quite a verb.  Maybe "Start a remote shell to log in..." or "Use a remote shell to log in..."

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+.Example output
+[source,terminal]
+----
+
```
Remove this blank line and the one on 147 just before the end of the code block.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
++
+[NOTE]
+====
+Depending on the resource that the user requests in the `pod-test.yaml` file, this can be any one of following three PCI addresses:
```
s/the user requests/you requested/

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
+====
+
+
+.  Change to the `test-bbdev` directory.
```
End the instructional text with a colon.  Consider mentioning that the directory is in the pod and not on our customers' computer.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
++
+[source,terminal]
+----
+sh-4.4# ./test-bbdev.py -e="-l ${CPU} -a ${PCIDEVICE_INTEL_COM_INTEL_FEC_5G}" -c validation \ -n 64 -b 32 -l 1 -v ./test_vectors/*"
```
Please make sure that you meant to include the backslash in the command.  Maybe the original text used that to add a newline and show the command on two lines.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,242 @@
++
+[NOTE]
+====
+While some tests can be skipped, be sure that the vector tests pass.
```
Unwrap this as a note and move it before the `.Example output` command or put a callout in the example output and use this text in the callout description.

mikemckiernan commented on scalability_and_performance/cnf-optimize-data-performance-n3000.adoc:
```diff
@@ -0,0 +1,26 @@
+
+include::modules/cnf-verify-sriov-operator-n3000.adoc[leveloffset=+2]
+
+[id="{context}-additional-resources"]
```
Team guidelines indicate to set "additional-resources_" as a prefix:

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies

mikemckiernan reviewed (changes requested):
There are still some requested changes outstanding.

kquinn1204 commented:
Got to leave this for min started changes this morning and will revisit later. Thanks for feedback @mikemckiernan

kquinn1204 commented:
Hi @mikemckiernan I think I have addressed most of these comments. I have asked about pods by label and I am waiting for a reply. Though SME has already given LGTM so I assume happy with the accuracy of commands

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,162 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="installing-the-openness-operator-intel-fpga-pac-n3000-operator_{context}"]
```
Please set the ID to match the file name according to the second item in the [module file metadata](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata) section of the doc guidelines.

mikemckiernan replied:
Please set the ID according to the guidelines in the preceding comment.

mikemckiernan replied:
Please set the ID as follows:

```
[id="installing-the-n3000-operator_{context}"]
```

If the ID cannot be set to that value--to meet the doc guidelines--please indicate the reason in a comment so that a subsequent reviewer does not request the same change.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,162 @@
+
+For vRAN use cases, the OpenNESS Operator for Intel FPGA PAC N3000 is used with the OpenNESS Operator for Wireless FEC Accelerators.
+
+As a cluster administrator, you can install the OpenNESS Operator for Intel® FPGA PAC N3000 by using the {product-title} CLI or the web console.
```
Please remove the reg symbol.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,162 @@
+$ oc create -f n3000-operatorgroup.yaml
+----
+
+
```
Please remove the blank line.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,162 @@
+$ oc create -f n3000-sub.yaml
+----
+
+. Verify that the OpenNESS Operator for Intel® FPGA PAC N3000 Operator is installed:
```
I apologize for failing to make an identical request that I made for the "fec-operator" file.  Please...

* Add a `.Verification` block title.
* Change the "." to "*" because it is a single-step procedure.
* Change the instructional text to "Verify the Operator is installed:" so it matches the fec-operator file.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,162 @@
+During installation, an Operator might display a *Failed* status. If the installation later succeeds with an *InstallSucceeded* message, you can ignore the *Failed* message.
+====
++
+If the Operator does not appear as installed, to troubleshoot further:
```
Please use the same (and better) text that you have in the fec-operator file:

"If the console does not indicate that the Operator is installed, perform the following troubleshooting steps:"

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="cnf-understanding-intel-hardware-accelertor-card-for-openshift-container-platform_{context}"]
```
Please match the ID to the file name as described in the second item of our [module file metadata](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#module-file-metadata) section of the doc guidelines.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+[id="cnf-understanding-intel-hardware-accelertor-card-for-openshift-container-platform_{context}"]
+= Understanding Intel hardware accelerator cards for OpenShift Container Platform
+
+Hardware accelerator cards from Intel accelerate 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads. This in turn increases in turn the overall compute capacity of a commercial off-the-shelf platform.
```
Please remove the second "in turn".

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+
+Hardware accelerator cards from Intel accelerate 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads. This in turn increases in turn the overall compute capacity of a commercial off-the-shelf platform.
+
+The hardware accelerator cards detailed in this section are:
```
Please remove this list because it is redundant.  If you believe customers need greater visibility, then consider changing the block titles to titles so that they appear in the TOC.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+
+.Intel FPGA PAC N3000
+
+The Intel FPGA PAC N3000 card needs flashed with the 4G/LTE or 5G bitstreams to support vRAN workloads.
```
"car needs flashed" is not grammatical.  Please revise.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+.Intel FPGA PAC N3000
+
+The Intel FPGA PAC N3000 card needs flashed with the 4G/LTE or 5G bitstreams to support vRAN workloads.
+The Intel FPGA PAC N3000 is a reference FPGA and uses 4G/LTE or 5G Forward Error Correction (FEC) as an example workload accelerating the 5G or 4G/LTE RAN layer 1 (L1) base station network function.
```
Please use "forward error correction" instead of the initial caps.  It is not a proper noun.

Please also make this into two sentences or change "accelerating" to "that accelerates".

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,39 @@
+
+Intel FPGA PAC N3000 support on {product-title} depends on two Operators:
+
+* OpenNESS Operator for Intel® FPGA PAC N3000 (Programming)
```
Please remove the reg symbol.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="cnf-programming-the-n3000-operator_{context}"]
```
This ID looks good. Thank you.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+
+When the Intel FPGA PAC N3000 is programmed with a vRAN 5G bitstream, the hardware exposes the Intel FPGA PAC N3000 with a vRAN 5G bitstream. This bitstream exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices used to accelerate the FEC in the vRAN workload.
+
+As a cluster administrator, you can install the OpenNESS Operator for Intel® FPGA PAC N3000 by using the {product-title} CLI or the web console.
```
Please remove the reg symbol.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+[id="cnf-programming-the-n3000-operator_{context}"]
+= Programming the OpenNESS Operator for Intel FPGA PAC N3000
+
+When the Intel FPGA PAC N3000 is programmed with a vRAN 5G bitstream, the hardware exposes the Intel FPGA PAC N3000 with a vRAN 5G bitstream. This bitstream exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices used to accelerate the FEC in the vRAN workload.
```
Please s/devices used to/devices that are used to/ so it reads a little more smoothly.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+[id="programmimg-n3000-with-a-vran-bitstream_{context}"]
+== Programming the N3000 with a vRAN bitstream
+
+As a cluster administrator, you can program the Intel FPGA PAC N3000 with a vRAN 5G bitstream. This bitstream exposes the Single Root I/O Virtualization (SR-IOV) virtual function (VF) devices used to accelerate the Forward Error Correction (FEC) in the vRAN workload.
```
Please lowercase "forward error correction" because it is not a proper noun.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+
+[NOTE]
+====
+The role of Forward Error Correction (FEC) is to correct transmission errors, where certain bits in a message can be lost or garbled. Messages can be lost or garbled due to noise in the transmission media, interference, or low signal strength.
```
Please remove the note markup.  This is the initial description of what FEC is.  It is not a noteworthy qualification, reminder, or anything else about FEC.

Pleas also lower case "forward error correction" because it is not a proper noun.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,343 @@
+----
++
+.Expected output
+
```
Please remove this blank line.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,350 @@
+[id="cnf-programming-the-sriov-operator-acc100_{context}"]
+= Configuring the SR-IOV-FEC Operator for the Intel vRAN Dedicated Accelerator ACC100
+
+Programming the Intel vRAN Dedicated Accelerator ACC100 exposes the Single Root I/O Virtualization (SRIOV) Virtual Function (VF) devices that are then used to accelerate the FEC in the vRAN workload.
```
Lowercase "virtual function" because it is not a proper noun.  Apologies for missing this earlier.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,350 @@
+= Configuring the SR-IOV-FEC Operator for the Intel vRAN Dedicated Accelerator ACC100
+
+Programming the Intel vRAN Dedicated Accelerator ACC100 exposes the Single Root I/O Virtualization (SRIOV) Virtual Function (VF) devices that are then used to accelerate the FEC in the vRAN workload.
+The Intel vRAN Dedicated Accelerator ACC100 accelerates 4G and 5G Virtualized Radio Access Networks (vRAN) workloads increasing the overall compute capacity of a commercial, off-the-shelf platform.
```
"workloads increasing" reads awkwardly.  Make it like the nearly identical sentence from `modules/cnf-intel-n3000-optimize-data-plane.adoc` after you remove the second "in turn."

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,350 @@
+The Intel vRAN Dedicated Accelerator ACC100 accelerates 4G and 5G Virtualized Radio Access Networks (vRAN) workloads increasing the overall compute capacity of a commercial, off-the-shelf platform.
+This device is also known as Mount Bryce.
+
+The SR-IOV-FEC Operator handles the management of the Forward Error Correction (FEC) devices used to accelerate the FEC process in vRAN L1 applications.
```
"forward error correction" because it is not a proper noun.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,350 @@
+* Configuring the VF queues for desired functionality in a 4G or 5G deployment
++
+[NOTE]
+====
```
Remove the NOTE markup.  Please see my similar comment in another review that indicates this is not noteworthy.

mikemckiernan commented on modules/cnf-programming-the-sriov-fec-operator-acc100.adoc:
```diff
@@ -0,0 +1,350 @@
+node1            Succeeded
+----
+
+. Find the physical function (PF) of the SR-IOV FEC accelerator device to be configured:
```
Please change "to be configured" to "to configure" for concision.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,403 @@
+= Configuring the SR-IOV-FEC Operator for Intel FPGA PAC N3000
+
+This section describes how to program the SR-IOV-FEC Operator for Intel FPGA PAC N3000.
+The SR-IOV-FEC Operator handles the management of the Forward Error Correction (FEC) devices used to accelerate the FEC process in vRAN L1 applications.
```
"forward error correction" because it is not a proper noun.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,403 @@
++
+[NOTE]
+====
+The role of Forward Error Correction (FEC) is to correct transmission errors, where certain bits in a message can be lost or garbled. Messages can be lost or garbled due to noise in the transmission media, interference, or low signal strength.
```
Please remove the NOTE markup.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,403 @@
+All the commands run in the `vran-acceleration-operators` namespace.
+====
+
+.Procedure
```
Please remove this redundant block title.  Please also consider removing the preceding note because it becomes superfluous after our customer runs the first command.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,403 @@
+node1            Succeeded
+----
+
+. Find the physical function (PF) of the SR-IOV FEC accelerator device to be configured:
```
Please change "to be configured" to "to configure" for concision.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,403 @@
+====
+For Intel PAC N3000 for vRAN Acceleration the user can create up to 8 VF devices. Each FEC PF device provides a total of 64 queues to be configured, 32 queues for uplink and 32 queues for downlink. The queues would be typically distributed evenly across the VFs.
+====
+.. Apply the CR:
```
Please add a blank line above this step.

If you have the time, please let me know which procedure you used as a touchstone--that uses "Apply the CR".  I'll put it on my list of "doc improvements" to make.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,237 @@
+//
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="cnf-verify-sriov-operator-n3000_{context}"]
```
ID looks good, thank you.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,237 @@
+[id="cnf-verify-sriov-operator-n3000_{context}"]
+= Verifying application pod access and FPGA usage on OpenNESS
+
+OpenNESS is an edge computing software toolkit toolkit that you can use to onboard and manage applications and network functions on any type of network.
```
Please remove duplicate "toolkit".

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,237 @@
+----
+sh-4.4#
+----
+. Print a list of environment variables:
```
Please add a blank line above this step.

mikemckiernan reviewed (commented):
Thanks for accepting a number of suggestions.  There are still a few remaining items to address.

kquinn1204 commented:
Hi @mikemckiernan I hope I picked up all these now. Thanks for your patience.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,163 @@
+[id="install-n3000operator-web-console_{context}"]
+== Installing the OpenNESS Operator for Intel FPGA PAC N3000 Operator by using the web console
+
+As a cluster administrator, you can install the OpenNESS Operator for Intel® FPGA PAC N3000 by using the web console.
```
Please remove the reg symbol.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,163 @@
+
+.Procedure
+
+. Install the OpenNESS Operator for Intel® FPGA PAC N3000 by using the {product-title} web console:
```
Please remove the reg symbol.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,163 @@
+
+.. In the {product-title} web console, click *Operators* -> *OperatorHub*.
+
+.. Choose *OpenNESS Operator for Intel® FPGA PAC N3000* from the list of available Operators, and then click *Install*.
```
Please remove the reg symbol, even if the symbol appears in the UI.

mikemckiernan commented on modules/cnf-installing-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,163 @@
+
+.. Switch to the *Operators* -> *Installed Operators* page.
+
+.. Ensure that *OpenNESS Operator for Intel® FPGA PAC N3000* is listed in the *vran-acceleration-operators* project with a *Status* of *InstallSucceeded*.
```
Please remove the reg symbol, even if it appears in the UI.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,32 @@
+
+Hardware accelerator cards from Intel accelerate 4G/LTE and 5G Virtualized Radio Access Networks (vRAN) workloads. This in turn increases the overall compute capacity of a commercial, off-the-shelf platform.
+
+.Intel FPGA PAC N3000
```
Please change this block title and the following block title to a regular title, such as "== Intel FPGA..." If you do not want the heading to appear in the TOC, you can set it as a [discrete heading](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#discrete-headings) as shown in the team doc guidelines.

mikemckiernan commented on modules/cnf-programming-the-n3000-operator.adoc:
```diff
@@ -0,0 +1,339 @@
+----
+n3000-daemonset-5k55l              1/1     Running   0          15d
+----
+.. View the logs:
```
Please add a line above this step to match the convention with other files in the repository.

mikemckiernan commented on modules/cnf-programming-the-sriov-operator.adoc:
```diff
@@ -0,0 +1,394 @@
+= Configuring the SR-IOV-FEC Operator for Intel FPGA PAC N3000
+
+This section describes how to program the SR-IOV-FEC Operator for Intel FPGA PAC N3000.
+The SR-IOV-FEC Operator handles the management of the Forward Error Correction (FEC) devices that are used to accelerate the FEC process in vRAN L1 applications.
```
lower case "forward error correction" because it is not a proper noun.

mikemckiernan commented on modules/cnf-verify-sriov-operator-n3000.adoc:
```diff
@@ -0,0 +1,238 @@
+* PCIDEVICE_INTEL_COM_INTEL_FEC_LTE
+
++
+.  Change to the `test-bbdev` directory:
```
I'm 99.99% sure that you should remove the `+` on line 155.  I can't imagine any purpose for it.

mikemckiernan reviewed (commented):
A few lingering items.  Please let me know if I can clarify my comments.

mikemckiernan commented on modules/cnf-intel-n3000-optimize-data-plane.adoc:
```diff
@@ -0,0 +1,36 @@
+// *cnf-optimize-data-performance-n3000.adoc
+
+[id="cnf-intel-n3000-optimize-data-plane_{context}"]
+= Understanding Intel hardware accelerator cards for OpenShift Container Platform
```
s/OpenShift Container Platform/{product-title}/

AFAIK, we can use attributes in titles and have several instances in the doc repository.

mikemckiernan reviewed (commented):
One or two more small comments.  Two highlights:

* In the verify-sriov-operator-n3000 file, there are several comments for a callout <4> that haven't been addressed or acknowledged.  If there is a reason why it should not be revised, can you provide it in a comment so that subsequent reviewers do not request a change?
* In that same file--possibly another--some requests to remove blank lines were not followed or acknowledged.  Removing blank lines seems like a worthwhile task to ease the writing task for subsequent writers in the same file.

kquinn1204 commented:
I have addresses those final comments. I think I had some spaces in my commands in the verify-sriov-operator-n3000 file which I have fixed. 
Hopefully good to go now. Thanks for taking the time to review

mikemckiernan reviewed (approved):
I still don't think this is English, "Hugepages and isolated CPUs needs to be configure using the Performance Addon Operator.", but it's been highlighted by two different reviewers on four different occasions.  I'm going to make believe it's me.
